### PR TITLE
feat(ci) : Import newman results to Xray

### PR DESCRIPTION
**Pull request by MedWassCh**
## Description

This PR is is dedicated to create test cases + test execution and add them into test plan and after that export newman results to the test execution

**Fixes** # (issue)

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [x] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software

## Target serie

- [x] 21.10.x
- [ ] 22.04.x
- [ ] 22.10.x
- [ ] 23.04.x
- [ ] 23.10.x (master)

<h2> How this pull request can be tested ? </h2>

Please describe the **procedure** to verify that the goal of the PR is matched. Provide clear instructions so that it can be **correctly tested**.

Any **relevant details** of the configuration to perform the test should be added.

#### Community contributors & Centreon team

- [ ] I have followed the **coding style guidelines** provided by Centreon
- [ ] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [ ] I have commented my code, especially **hard-to-understand areas** of the PR.
- [ ] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

File: .github/workflows/newman.yml [hunk 1]
```diff
@@ -21,7 +21,14 @@ on:
...
         required: true
-
+      CLIENT_ID :
+        required: true
+      CLIENT_SECRET:
+        required: true
+      JIRA_USER:
+          required: true
+      JIRA_TOKEN_TEST:
+          required: true
```
Comment (kduret) [inline .github/workflows/newman.yml:31]: for consistency with other resuable workflow, use secret in lowercase
```suggestion
      client_id:
        required: true
      client_secret:
        required: true
      jira_user:
          required: true
      jira_token_test:
          required: true

```

File: .github/workflows/newman.yml [hunk 2]
```diff
@@ -41,8 +48,179 @@ jobs:
           echo "GITHUB_OUTPUT contents:"
           cat $GITHUB_OUTPUT
 
+  associate-test-cases:
```
Comment (kduret) [inline .github/workflows/newman.yml:51]: as discussed, this workflow should be run in following environments : `unstable` , `testing` , `stable`

File: .github/workflows/newman.yml [hunk 3]
```diff
@@ -41,8 +48,179 @@ jobs:
...
+      steps:
+        - name: Checkout Repository
+          uses: actions/checkout@v4
+
+        - name: Generate Xray Token
+          id: generate-xray-token
+          run: |
+                  token_response=$(curl -H "Content-Type: application/json" -X POST --data "{\"client_id\": \"${{ secrets.CLIENT_ID }}\", \"client_secret\": \"${{ secrets.CLIENT_SECRET }}\"}" "https://xray.cloud.getxray.app/api/v1/authenticate")
+                  xray_token=$(echo "$token_response" | sed -n 's/.*"\(.*\)".*/\1/p')
+                  echo "this is the xray token : $xray_token"
```
Comment (kduret) [inline .github/workflows/newman.yml:70]: do not display token for security reasons
```suggestion
```

File: .github/workflows/newman.yml [hunk 4]
```diff
@@ -41,8 +48,179 @@ jobs:
...
+                  echo "xray_token=$xray_token" >> $GITHUB_OUTPUT
+          shell: bash
+
+        - name: Associate Test Cases with Test plan
+          id: xray-newman
+          run: |
+            xray_graphql_getTestPlan='{
+              "query": "query GetTestPlan($issueId: String) { getTestPlan(issueId: $issueId) { issueId tests(limit: 100) { results { issueId testType { name } } } } }",
+              "variables": {
+                "issueId": "73560"
```
Comment (kduret) [inline .github/workflows/newman.yml:80]: wait current @a-eljazouly work to get this value dynamically or create it if it does not exist
@a-eljazouly maybe do a separated action which generate an output `testPlanId` to reuse it in this workflow

File: .github/workflows/newman.yml [hunk 5]
```diff
@@ -41,8 +48,179 @@ jobs:
...
+              jira_issue_url="https://centreon.atlassian.net/rest/api/2/issue/$issue_id"
+
+              response=$(curl --request GET \
+                --url "$jira_issue_url" \
+                --user "${{ secrets.JIRA_USER }}:${{ secrets.JIRA_TOKEN_TEST }}" \
+                --header 'Accept: application/json')
+
+              summary=$(echo "$response" | jq -r '.fields.summary')
+
+              if [ "$(echo "$response" | jq -r '.errorMessages[0]')" == "Issue does not exist or you do not have permission to see it." ]; then
```
Comment (kduret) [inline .github/workflows/newman.yml:104]: Use response code instead of error message for robustness

File: .github/workflows/newman.yml (hunk 4 above)
Comment (kduret) [reply .github/workflows/newman.yml:80]: @MedWassCh just use an other id regarding environments

File: .github/workflows/newman.yml [hunk 6]
```diff
@@ -41,8 +48,179 @@ jobs:
...
+              }
+            }'
+            for collection_file in "${collections[@]}"; do
+              collection_name=$(basename "$collection_file" .postman_collection.json)
+              collection_name_sanitized="${collection_name//[^a-zA-Z0-9]/_}"
+
+              if [[ " ${summaries[@]} " =~ " ${collection_name_sanitized} " ]]; then
+                echo "The test case for $collection_name_sanitized already exists in the test plan."
+              else
+                # Ajouter un cas de test
```
Comment (kduret) [inline .github/workflows/newman.yml:130]: add comments in english

File: .github/workflows/newman.yml [hunk 7]
```diff
@@ -41,8 +48,179 @@ jobs:
...
+                sleep 5
+
+                if [ -z "$response" ]; then
+                  echo "Failed to create the test case within the specified time."
+                else
+                  test_case_id=$(echo "$response" | jq -r '.id')
+                  test_case_ids+=("$test_case_id")  # Ajouter l'ID du cas de test à la liste
+
+                  echo "Created Test Case with ID: $test_case_id"
+                  # Ajouter le nom du cas de test à la liste des résumés
```
Comment (kduret) [inline .github/workflows/newman.yml:159]: add comments in english

File: .github/workflows/newman.yml [hunk 8]
```diff
@@ -41,8 +48,179 @@ jobs:
...
+                      "summary": "'"$collection_name_sanitized"'",
+                      "description": "Test case for '"$collection_name_sanitized"'",
+                      "issuetype": {
+                        "name": "Test"
+                      }
+                    }
+                  }' \
+                  --max-time 20)
+
+                sleep 5
```
Comment (kduret) [inline .github/workflows/newman.yml:150]: why do you need to wait ?

File: .github/workflows/newman.yml [hunk 9]
```diff
@@ -41,8 +48,179 @@ jobs:
...
+              echo "Failed to create the Test Execution. Check the response for errors."
+            else
+              echo "Created Test Execution with ID: $test_execution_id"
+              echo "test_exec=$test_execution_id" >> $GITHUB_OUTPUT
+              fi
+
+            xray_graphql_addTestExecutionsToTestPlan='{
+              "query": "mutation AddTestExecutionsToTestPlan($issueId: String!, $testExecIssueIds: [String]!) { addTestExecutionsToTestPlan(issueId: $issueId, testExecIssueIds: $testExecIssueIds) { addedTestExecutions warning } }",
+              "variables": {
+                "issueId": "73560",
```
Comment (kduret) [inline .github/workflows/newman.yml:216]: use id stored in a variable

File: .github/workflows/newman.yml [hunk 10]
```diff
@@ -41,8 +48,179 @@ jobs:
...
+              fi
+            done
+
+            collections=($(find ./collections -type f -name "*.postman_collection.json"))
+            test_case_ids=()
+
+            xray_graphql_AddingTestsToTestPlan='{
+              "query": "mutation AddTestsToTestPlan($issueId: String!, $testIssueIds: [String]!) { addTestsToTestPlan(issueId: $issueId, testIssueIds: $testIssueIds) { addedTests warning } }",
+              "variables": {
+                "issueId": "73560",
```
Comment (kduret) [inline .github/workflows/newman.yml:119]: use id stored in a variable

File: .github/workflows/newman.yml [hunk 11]
```diff
@@ -98,28 +280,32 @@ jobs:
     - name: Wait for Centreon Web Container to Start
       run: pnpx wait-on http://localhost:4000/ --timeout 100000
 
+    - name: Generate Xray Token
+      id: generate-xray-token
+      run: |
+        token_response=$(curl -H "Content-Type: application/json" -X POST --data "{\"client_id\": \"${{ secrets.CLIENT_ID }}\", \"client_secret\": \"${{ secrets.CLIENT_SECRET }}\"}" "https://xray.cloud.getxray.app/api/v1/authenticate")
+        xray_token=$(echo "$token_response" | sed -n 's/.*"\(.*\)".*/\1/p')
+        echo "this is the xray token : $xray_token"
```
Comment (kduret) [inline .github/workflows/newman.yml:288]:
```suggestion
```

File: .github/workflows/newman.yml [hunk 12]
```diff
@@ -134,6 +320,90 @@ jobs:
...
 
+    - name: Change test execution status
+      run: |
+        for summary_file in ./postman_summaries/*.json; do
+             test_execution_id=${{ needs.associate-test-cases.outputs.test_exec }}
+            filename=$(basename "$summary_file")
+            collection_name="${filename%-summary.json}"
+            collection_name="${collection_name//-/_}"
+            echo "Contenu de collection_name : $collection_name"
+            failed_requests=$(jq -r '.Run.Stats.Requests.failed' "$summary_file")
```
Comment (kduret) [inline .github/workflows/newman.yml:331]: use 2 spaces indentation

File: .github/workflows/newman.yml [hunk 13]
```diff
@@ -134,6 +320,90 @@ jobs:
...
+                if [ "$summary" == "$collection_name" ]; then
+                  echo "Matching issue ID: $issue_id with Collection Name: $collection_name"
+                  if [ "$failed_requests" -gt 0 ] || [ "$failed_assertions" -gt 0 ]; then
+                    newStatus="FAIL"
+                  else
+                    newStatus="PASSED"
+                  fi
+                  # Create a JSON array from IDs
+                  testIssueIds=($issue_id)
+                  echo "Contenu de testIssueIds : ${testIssueIds[@]}"
```
Comment (kduret) [inline .github/workflows/newman.yml:371]: translate in english

File: .github/workflows/newman.yml [hunk 14]
```diff
@@ -134,6 +320,90 @@ jobs:
...
+            # Parsing and processing tests id
+            issue_ids=($(echo "$response" | jq -r '.data.getTestPlan.tests.results[].issueId'))
+            summaries=()
+            for issue_id in "${issue_ids[@]}"; do
+              echo "Processing issue ID: $issue_id"
+              jira_issue_url="https://centreon.atlassian.net/rest/api/2/issue/$issue_id"
+
+              response=$(curl --request GET --url "$jira_issue_url" --user "${{ secrets.JIRA_USER }}:${{ secrets.JIRA_TOKEN_TEST }}" --header 'Accept: application/json')
+
+              if [ "$(echo "$response" | jq -r '.errorMessages[0]')" == "Issue does not exist or you do not have permission to see it." ]; then
```
Comment (kduret) [inline .github/workflows/newman.yml:355]: use response code instead of error message for robustness

File: .github/workflows/web.yml [hunk 15]
```diff
@@ -504,6 +504,10 @@ jobs:
     secrets:
       registry_username: ${{ secrets.DOCKER_REGISTRY_ID}}
       registry_password: ${{ secrets.DOCKER_REGISTRY_PASSWD }}
+      CLIENT_ID: ${{ secrets.CLIENT_ID }}
+      CLIENT_SECRET : ${{ secrets.CLIENT_SECRET }}
+      JIRA_USER: ${{ secrets.JIRA_USER }}
+      JIRA_TOKEN_TEST: ${{ secrets.JIRA_TOKEN_TEST }}
```
Comment (kduret) [inline .github/workflows/web.yml:510]:
```suggestion
      client_id: ${{ secrets.CLIENT_ID }}
      client_secret: ${{ secrets.CLIENT_SECRET }}
      jira_user: ${{ secrets.JIRA_USER }}
      jira_token_test: ${{ secrets.JIRA_TOKEN_TEST }}
```

File: centreon/tests/rest_api/package.json [hunk 16]
```diff
@@ -8,6 +8,8 @@
     "newman": "^5.3.2",
     "newman-reporter-htmlextra": "^1.22.11",
     "newman-reporter-json-summary": "^1.0.14",
+    "newman-reporter-junitfull": "^1.1.1",
+    "newman-reporter-junitxray": "^1.0.3",
```
Comment (kduret) [inline centreon/tests/rest_api/package.json:12]: bot are used ?

File: .github/workflows/newman.yml [hunk 17]
```diff
@@ -21,7 +21,14 @@ on:
         required: true
       registry_password:
         required: true
-
+      client_id :
```
Comment (kduret) [inline .github/workflows/newman.yml:24]:
```suggestion
      client_id:
```

File: .github/workflows/newman.yml [hunk 18]
```diff
@@ -98,28 +279,37 @@ jobs:
...
+        fi
 
     - name: Upload HTML Reports
       if: failure()
       uses: actions/upload-artifact@v3
       with:
-        name: postman-html-reports
-        path: centreon/tests/rest_api/newman/
+          name: postman-html-reports
+          path: centreon/tests/rest_api/newman/
```
Comment (kduret) [inline .github/workflows/newman.yml:310]:
```suggestion
        name: postman-html-reports
        path: centreon/tests/rest_api/newman/

```

File: .github/workflows/web.yml [hunk 19]
```diff
@@ -504,6 +504,10 @@ jobs:
     secrets:
       registry_username: ${{ secrets.DOCKER_REGISTRY_ID}}
       registry_password: ${{ secrets.DOCKER_REGISTRY_PASSWD }}
+      CLIENT_ID: ${{ secrets.client_id }}
+      CLIENT_SECRET : ${{ secrets.client_secret }}
+      JIRA_USER: ${{ secrets.jira_user }}
+      JIRA_TOKEN_TEST: ${{ secrets.jira_token_test }}
```
Comment (kduret) [inline .github/workflows/web.yml:510]:
```suggestion
      client_id: ${{ secrets.client_id }}
      client_secret: ${{ secrets.client_secret }}
      jira_user: ${{ secrets.jira_user }}
      jira_token_test: ${{ secrets.jira_token_test }}
```

File: .github/workflows/newman.yml [hunk 20]
```diff
@@ -41,8 +54,219 @@ jobs:
           echo "GITHUB_OUTPUT contents:"
           cat $GITHUB_OUTPUT
 
+  associate-test-cases:
+      runs-on: ubuntu-22.04
+      outputs:
+        test_exec: ${{ steps.xray-newman.outputs.test_exec }}
+        test_plan_id: ${{ steps.get-testPlan-id.outputs.testPlan_id }}
```
Comment (kduret) [inline .github/workflows/newman.yml:61]: update variables for more consistency
```suggestion
        test_plan_id: ${{ steps.get-test-plan-id.outputs.test_plan_id }}
```

File: .github/workflows/newman.yml [hunk 21]
```diff
@@ -41,8 +54,219 @@ jobs:
           echo "GITHUB_OUTPUT contents:"
           cat $GITHUB_OUTPUT
 
+  associate-test-cases:
+      runs-on: ubuntu-22.04
+      outputs:
+        test_exec: ${{ steps.xray-newman.outputs.test_exec }}
```
Comment (kduret) [inline .github/workflows/newman.yml:60]:
```suggestion
        test_execution_id: ${{ steps.xray-newman.outputs.test_execution_id }}
```

File: .github/workflows/newman.yml [hunk 22]
```diff
@@ -41,8 +54,219 @@ jobs:
...
+
+              response=$(curl --request GET \
+                --url "$jira_issue_url" \
+                --user "${{ secrets.jira_user }}:${{ secrets.jira_token_test }}" \
+                --header 'Accept: application/json')
+
+              summary=$(echo "$response" | jq -r '.fields.summary')
+
+              if [ "$response_code" -eq 404 ]; then
+               echo "The issue with ID $issue_id does not exist or you do not have permission to see it."
```
Comment (kduret) [inline .github/workflows/newman.yml:150]:
```suggestion
                echo "The issue with ID $issue_id does not exist or you do not have permission to see it."
```

File: .github/workflows/newman.yml [hunk 23]
```diff
@@ -98,18 +321,27 @@ jobs:
...
       run: |
         collection_file="collections/${{ matrix.feature }}"
         collection_name=$(basename "$collection_file" .postman_collection.json)
         collection_directory="$(dirname "$collection_file")"
         environment_file=$(find "$collection_directory" -maxdepth 1 -type f -name "*.postman_environment.json")
-          if [ -f "$environment_file" ]; then
-            pnpm newman run "$collection_file" -e "$environment_file" --reporters cli,htmlextra,json-summary --reporter-htmlextra-title "$collection_name" --reporter-htmlextra-title "${collection_name// /_}" --reporter-htmlextra-export "newman/${collection_name}.html" --reporter-summary-json-export "postman_summaries/${collection_name}-summary.json"
+        if [ -f "$environment_file" ]; then
+          echo "Environment: $environment_file"
+           pnpm newman run "$collection_file" -e "$environment_file" --reporters cli,htmlextra,json-summary --reporter-htmlextra-title "$collection_name" --reporter-htmlextra-title "${collection_name// /_}" --reporter-htmlextra-export "newman/${collection_name}.html" --reporter-summary-json-export "postman_summaries/${collection_name}-summary.json"
```
Comment (kduret) [inline .github/workflows/newman.yml:340]:
```suggestion
          pnpm newman run "$collection_file" -e "$environment_file" --reporters cli,htmlextra,json-summary --reporter-htmlextra-title "$collection_name" --reporter-htmlextra-title "${collection_name// /_}" --reporter-htmlextra-export "newman/${collection_name}.html" --reporter-summary-json-export "postman_summaries/${collection_name}-summary.json"
```

File: .github/workflows/newman.yml [hunk 24]
```diff
@@ -98,18 +321,27 @@ jobs:
...
           else
-            echo "The file Environment-Collection.json was not found in the same directory as $collection_file"
-            exit 1
-          fi
+          echo "The file Environment-Collection.json was not found in the same directory as $collection_file"
+          exit 1
+        fi
```
Comment (kduret) [inline .github/workflows/newman.yml:344]:
```suggestion
            echo "The file Environment-Collection.json was not found in the same directory as $collection_file"
            exit 1
          fi
```

**Review by kduret — COMMENTED** [review]
I guess `test` and `test copy` directories must be removed

**Review by kduret — APPROVED** [review]

**Review by paul-oureib — APPROVED** [review]

**Pull request merged**